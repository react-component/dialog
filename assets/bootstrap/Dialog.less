@import "./variables.less";

.clearfix() {
  &:before,
  &:after {
    content: " "; // 1
    display: table; // 2
  }
  &:after {
    clear: both;
  }
}

.@{prefixCls} {
  // Container that the rc-dialog scrolls within
  &-wrap {
    position: fixed;
    overflow: auto;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    z-index: @zindex-modal;
    -webkit-overflow-scrolling: touch;

    // Prevent Chrome on Windows from adding a focus outline. For details, see
    // https://github.com/twbs/bootstrap/pull/10951.
    outline: 0;
  }

  // Shell div to position the rc-dialog with bottom padding
  position: relative;
  width: auto;
  margin: 10px;

  // Actual rc-dialog
  &-section {
    position: relative;
    background-color: @modal-section-bg;
    border: 1px solid @modal-section-fallback-border-color; //old browsers fallback (ie8 etc)
    border: 1px solid @modal-section-border-color;
    border-radius: @border-radius-large;
    box-shadow: 0 3px 9px rgba(0, 0, 0, .5);
    background-clip: padding-box;
    // Remove focus outline from opened rc-dialog
    outline: 0;
  }

  // Modal header
  // Top section of the rc-dialog w/ title and dismiss
  &-header {
    padding: @modal-title-padding;
    border-bottom: 1px solid @modal-header-border-color;
    &:extend(.clearfix all);
  }

  &-close {
    cursor: pointer;
    border: 0;
    background: transparent;
    font-size: 21px;
    position: absolute;
    right: 20px;
    top: 12px;
    font-weight: 700;
    line-height: 1;
    color: #000;
    text-shadow: 0 1px 0 #fff;
    filter: alpha(opacity=20);
    opacity: .2;
    text-decoration: none;

    &-x:after {
      content: 'Ã—'
    }

    &:hover {
      opacity: 1;
      filter: alpha(opacity=100);
      text-decoration: none;
    }
  }

  // Title text within header
  &-title {
    margin: 0;
    line-height: @modal-title-line-height;
  }

  // Modal body
  // Where all rc-dialog content resides (sibling of &-header and &-footer)
  &-body {
    position: relative;
    padding: @modal-inner-padding;
  }

  // Footer (for actions)
  &-footer {
    padding: @modal-inner-padding;
    text-align: right; // right align buttons
    border-top: 1px solid @modal-footer-border-color;
    &:extend(.clearfix all); // clear it in case folks use .pull-* classes on buttons

    // Properly space out buttons
    .btn + .btn {
      margin-left: 5px;
      margin-bottom: 0; // account for input[type="submit"] which gets the bottom margin like all other inputs
    }
    // but override that for button groups
    .btn-group .btn + .btn {
      margin-left: -1px;
    }
    // and override it for block buttons as well
    .btn-block + .btn-block {
      margin-left: 0;
    }
  }
}

// Scale up the rc-dialog
@media (min-width: @screen-sm-min) {
  .@{prefixCls} {
    // Automatically set rc-dialog's width for larger viewports
    width: @modal-md;
    margin: 30px auto;

    &-section {
      box-shadow: 0 5px 15px rgba(0, 0, 0, .5);
    }
  }
}

@import "./effect.less";
