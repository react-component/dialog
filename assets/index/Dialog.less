.@{prefixCls} {
  position: relative;
  width: auto;
  margin: 10px;

  &-wrap {
    position: fixed;
    overflow: auto;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    z-index: 1050;
    -webkit-overflow-scrolling: touch;
    outline: 0;
  }

  &-title {
    margin: 0;
    font-size: 14px;
    line-height: 21px;
    font-weight: bold;
  }

  &-section {
    position: relative;
    background-color: #ffffff;
    border: none;
    border-radius: 6px 6px;
    background-clip: padding-box;
  }

  &-close {
    cursor: pointer;
    border: 0;
    background: transparent;
    font-size: 21px;
    position: absolute;
    right: 20px;
    top: 12px;
    font-weight: 700;
    line-height: 1;
    color: #000;
    text-shadow: 0 1px 0 #fff;
    filter: alpha(opacity=20);
    opacity: .2;
    text-decoration: none;

    &:disabled {
      pointer-events: none;
    }

    &-x:after {
      content: 'Ã—'
    }

    &:hover {
      opacity: 1;
      filter: alpha(opacity=100);
      text-decoration: none;
    }
  }

  &-header {
    padding: 13px 20px 14px 20px;
    border-radius: 5px 5px 0 0;
    background: #fff;
    color: #666;
    border-bottom: 1px solid #e9e9e9;
  }

  &-body {
    padding: 20px;
  }

  &-footer {
    border-top: 1px solid #e9e9e9;
    padding: 10px 20px;
    text-align: right;
    border-radius: 0 0 5px 5px;
  }

  .effect() {
    animation-duration: 0.3s;
    animation-fill-mode: both;
  }

  &-zoom-enter, &-zoom-appear {
    opacity: 0;
    .effect();
    animation-timing-function: cubic-bezier(0.08, 0.82, 0.17, 1);
    animation-play-state: paused;
  }

  &-zoom-leave {
    .effect();
    animation-timing-function: cubic-bezier(0.6, 0.04, 0.98, 0.34);
    animation-play-state: paused;
  }

  &-zoom-enter&-zoom-enter-active, &-zoom-appear&-zoom-appear-active {
    animation-name: rcDialogZoomIn;
    animation-play-state: running;
  }

  &-zoom-leave&-zoom-leave-active {
    animation-name: rcDialogZoomOut;
    animation-play-state: running;
  }

  @keyframes rcDialogZoomIn {
    0% {
      opacity: 0;
      transform: scale(0, 0);
    }
    100% {
      opacity: 1;
      transform: scale(1, 1);
    }
  }
  @keyframes rcDialogZoomOut {
    0% {

      transform: scale(1, 1);
    }
    100% {
      opacity: 0;
      transform: scale(0, 0);
    }
  }
}

@media (min-width: 768px) {
  .@{prefixCls} {
    width: 600px;
    margin: 30px auto;
  }
}
