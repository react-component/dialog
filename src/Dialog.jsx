import React, { PropTypes } from 'react';
import ReactDOM from 'react-dom';
import KeyCode from 'rc-util/lib/KeyCode';
import Animate from 'rc-animate';
import LazyRenderBox from './LazyRenderBox';
let uuid = 0;
let openCount = 0;

function noop() {
}

function getScroll(w, top) {
  let ret = w[`page${top ? 'Y' : 'X'}Offset`];
  const method = `scroll${top ? 'Top' : 'Left'}`;
  if (typeof ret !== 'number') {
    const d = w.document;
    ret = d.documentElement[method];
    if (typeof ret !== 'number') {
      ret = d.body[method];
    }
  }
  return ret;
}

function setTransformOrigin(node, value) {
  const style = node.style;
  ['Webkit', 'Moz', 'Ms', 'ms'].forEach((prefix) => {
    style[`${prefix}TransformOrigin`] = value;
  });
  style[`transformOrigin`] = value;
}

function offset(el) {
  const rect = el.getBoundingClientRect();
  const pos = {
    left: rect.left,
    top: rect.top,
  };
  const doc = el.ownerDocument;
  const w = doc.defaultView || doc.parentWindow;
  pos.left += getScroll(w);
  pos.top += getScroll(w, 1);
  return pos;
}

const Dialog = React.createClass({
  propTypes: {
    onAfterClose: PropTypes.func,
    onClose: PropTypes.func,
    closable: PropTypes.bool,
    maskClosable: PropTypes.bool,
    visible: PropTypes.bool,
    mousePosition: PropTypes.object,
    wrapStyle: PropTypes.object,
    wrapClassName: PropTypes.string,
  },

  getDefaultProps() {
    return {
      onAfterClose: noop,
      onClose: noop,
    };
  },

  componentWillMount() {
    this.titleId = `rcDialogTitle${uuid++}`;
  },

  componentDidMount() {
    this.componentDidUpdate({});
  },

  componentDidUpdate(prevProps) {
    const props = this.props;
    const mousePosition = this.props.mousePosition;
    if (props.visible) {
      // first show
      if (!prevProps.visible) {
        this.lastOutSideFocusNode = document.activeElement;
        this.addScrollingClass();
        this.refs.wrap.focus();
        const dialogNode = ReactDOM.findDOMNode(this.refs.dialog);
        if (mousePosition) {
          const elOffset = offset(dialogNode);
          setTransformOrigin(dialogNode,
            `${mousePosition.x - elOffset.left}px ${mousePosition.y - elOffset.top}px`);
        } else {
          setTransformOrigin(dialogNode, '');
        }
      }
    } else if (prevProps.visible) {
      if (props.mask && this.lastOutSideFocusNode) {
        try {
          this.lastOutSideFocusNode.focus();
        } catch (e) {
          this.lastOutSideFocusNode = null;
        }
        this.lastOutSideFocusNode = null;
        this.removeScrollingClass();
      }
    }
  },

  onAnimateLeave() {
    if (ReactDOM.findDOMNode(this.refs.wrap)) {
      this.refs.wrap.style.display = 'none';
    }
    this.props.onAfterClose();
  },

  onMaskClick(e) {
    if (e.target === e.currentTarget && this.props.closable && this.props.maskClosable) {
      this.close(e);
    }
  },

  onKeyDown(e) {
    const props = this.props;
    if (props.closable && props.keyboard) {
      if (e.keyCode === KeyCode.ESC) {
        this.close(e);
      }
    }
    // keep focus inside dialog
    if (props.visible) {
      if (e.keyCode === KeyCode.TAB) {
        const activeElement = document.activeElement;
        const dialogRoot = this.refs.wrap;
        const sentinel = this.refs.sentinel;
        if (e.shiftKey) {
          if (activeElement === dialogRoot) {
            sentinel.focus();
          }
        } else if (activeElement === this.refs.sentinel) {
          dialogRoot.focus();
        }
      }
    }
  },

  getDialogElement() {
    const props = this.props;
    const closable = props.closable;
    const prefixCls = props.prefixCls;
    const dest = {};
    if (props.width !== undefined) {
      dest.width = props.width;
    }
    if (props.height !== undefined) {
      dest.height = props.height;
    }


    let footer;
    if (props.footer) {
      footer = (<div className={`${prefixCls}-footer`} ref="footer">
        {props.footer}
      </div>);
    }

    let header;
    if (props.title) {
      header = (<div className={`${prefixCls}-header`} ref="header">
        <div className={`${prefixCls}-title`} id={this.titleId}>
          {props.title}
        </div>
      </div>);
    }

    let closer;
    if (closable) {
      closer = (<button
        onClick={this.close}
        aria-label="Close"
        className={`${prefixCls}-close`}
      >
        <span className={`${prefixCls}-close-x`}/>
      </button>);
    }

    const style = {
      ...props.style,
      ...dest,
    };
    const transitionName = this.getTransitionName();
    const dialogElement = (
      <LazyRenderBox
        role="document"
        ref="dialog"
        style={style}
        className={`${prefixCls} ${props.className || ''}`}
        visible={props.visible}
      >
        <div className={`${prefixCls}-content`}>
          {closer}
          {header}
          <div className={`${prefixCls}-body`} style={props.bodyStyle} ref="body">
            {props.children}
          </div>
          {footer}
        </div>
        <div tabIndex="0" ref="sentinel" style={{ width: 0, height: 0, overflow: 'hidden' }}>
          sentinel
        </div>
      </LazyRenderBox>
    );
    return (
      <Animate
        key="dialog"
        showProp="visible"
        onLeave={this.onAnimateLeave}
        transitionName={transitionName}
        component=""
        transitionAppear
      >
        {dialogElement}
      </Animate>
    );
  },

  getZIndexStyle() {
    const style = {};
    const props = this.props;
    if (props.zIndex !== undefined) {
      style.zIndex = props.zIndex;
    }
    return style;
  },

  getWrapStyle() {
    return {
      ...this.getZIndexStyle(),
      ...this.props.wrapStyle,
    };
  },

  getMaskElement() {
    const props = this.props;
    let maskElement;
    if (props.mask) {
      const maskTransition = this.getMaskTransitionName();
      maskElement = (
        <LazyRenderBox
          style={this.getZIndexStyle()}
          key="mask"
          className={`${props.prefixCls}-mask`}
          hiddenClassName={`${props.prefixCls}-mask-hidden`}
          visible={props.visible}
        />
      );
      if (maskTransition) {
        maskElement = (
          <Animate
            key="mask"
            showProp="visible"
            transitionAppear
            component=""
            transitionName={maskTransition}
          >
            {maskElement}
          </Animate>
        );
      }
    }
    return maskElement;
  },

  getMaskTransitionName() {
    const props = this.props;
    let transitionName = props.maskTransitionName;
    const animation = props.maskAnimation;
    if (!transitionName && animation) {
      transitionName = `${props.prefixCls}-${animation}`;
    }
    return transitionName;
  },

  getTransitionName() {
    const props = this.props;
    let transitionName = props.transitionName;
    const animation = props.animation;
    if (!transitionName && animation) {
      transitionName = `${props.prefixCls}-${animation}`;
    }
    return transitionName;
  },

  getElement(part) {
    return this.refs[part];
  },

  addScrollingClass() {
    openCount++;
    if (openCount !== 1) {
      return;
    }
    const props = this.props;
    const scrollingClassName = `${props.prefixCls}-open`;
    document.body.className += ` ${scrollingClassName}`;
  },

  removeScrollingClass() {
    openCount--;
    if (openCount !== 0) {
      return;
    }
    const props = this.props;
    const scrollingClassName = `${props.prefixCls}-open`;
    const body = document.body;
    body.className = body.className.replace(scrollingClassName, '');
  },

  close(e) {
    this.props.onClose(e);
  },

  render() {
    const props = this.props;
    const prefixCls = props.prefixCls;
    const style = this.getWrapStyle();
    // clear hide display
    // and only set display after async anim, not here for hide
    if (props.visible) {
      style.display = null;
    }
    return (<div>
      {this.getMaskElement()}
      <div
        tabIndex="-1"
        onKeyDown={this.onKeyDown}
        className={`${prefixCls}-wrap ${props.wrapClassName || ''}`}
        ref="wrap"
        onClick={this.onMaskClick}
        role="dialog"
        aria-labelledby={props.title ? this.titleId : null}
        style={style}
      >
        {this.getDialogElement()}
      </div>
    </div>);
  },
});

export default Dialog;
