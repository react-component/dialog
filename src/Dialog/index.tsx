import * as React from 'react';
import { useRef, useEffect } from 'react';
import classNames from 'classnames';
import KeyCode from 'rc-util/lib/KeyCode';
import contains from 'rc-util/lib/Dom/contains';
import { IDialogPropTypes } from '../IDialogPropTypes';
import Mask from './Mask';
import { getMotionName, getUUID } from '../util';
import Content, { ContentRef } from './Content';

export interface IDialogChildProps extends IDialogPropTypes {
  // zombieJ: This should be handle on top instead of each Dialog.
  // TODO: refactor to remove this.
  getOpenCount: () => number;
  switchScrollingEffect?: () => void;
}

export default function Dialog(props: IDialogChildProps) {
  const {
    prefixCls = 'rc-dialog',
    zIndex,
    visible = false,
    keyboard = true,
    focusTriggerAfterClose = true,
    switchScrollingEffect = () => {},

    // Wrapper
    title,
    wrapStyle,
    wrapClassName,
    wrapProps,
    onClose,
    afterClose,

    // Dialog
    transitionName,
    animation,
    closable = true,

    // Mask
    mask = true,
    maskTransitionName,
    maskAnimation,
    maskClosable = true,
    maskStyle,
    maskProps,
  } = props;

  const lastOutSideActiveElementRef = useRef<HTMLElement>();
  const wrapperRef = useRef<HTMLDivElement>();
  const contentRef = useRef<ContentRef>();

  const [animatedVisible, setAnimatedVisible] = React.useState(visible);

  // ========================== Init ==========================
  const ariaIdRef = useRef<string>();
  if (!ariaIdRef.current) {
    ariaIdRef.current = `rcDialogTitle${getUUID()}`;
  }

  // ========================= Events =========================
  function onDialogVisibleChanged(newVisible: boolean) {
    if (newVisible) {
      // Try to focus
      if (!contains(wrapperRef.current, document.activeElement)) {
        lastOutSideActiveElementRef.current = document.activeElement as HTMLElement;
        contentRef.current?.focus();
      }
    } else {
      // Clean up scroll bar & focus back
      setAnimatedVisible(false);
      switchScrollingEffect();

      if (mask && lastOutSideActiveElementRef.current && focusTriggerAfterClose) {
        try {
          lastOutSideActiveElementRef.current.focus({ preventScroll: true });
        } catch (e) {
          // Do nothing
        }
        lastOutSideActiveElementRef.current = null;
      }

      afterClose?.();
    }
  }

  function onInternalClose(e: React.SyntheticEvent) {
    onClose?.(e);
  }

  // >>> Content
  const contentClickRef = useRef(false);
  const contentTimeoutRef = useRef<number>();

  // We need record content click incase content popup out of dialog
  const onContentClick: React.MouseEventHandler = () => {
    clearTimeout(contentTimeoutRef.current);
    contentClickRef.current = true;

    contentTimeoutRef.current = setTimeout(() => {
      contentClickRef.current = false;
    });
  };

  // >>> Wrapper
  // Close only when element not on dialog
  let onWrapperClick: (e: React.SyntheticEvent) => void = null;
  if (maskClosable) {
    onWrapperClick = (e) => {
      if (
        !contentClickRef.current &&
        !contains(contentRef.current.getDOM(), e.target as HTMLElement)
      ) {
        onInternalClose(e);
      }
    };
  }

  function onWrapperKeyDown(e: React.KeyboardEvent<HTMLDivElement>) {
    if (keyboard && e.keyCode === KeyCode.ESC) {
      e.stopPropagation();
      onInternalClose(e);
      return;
    }

    // keep focus inside dialog
    if (visible) {
      if (e.keyCode === KeyCode.TAB) {
        contentRef.current.changeActive(!e.shiftKey);
      }
    }
  }

  // ========================= Effect =========================
  useEffect(() => {
    if (visible) {
      setAnimatedVisible(true);
      switchScrollingEffect();
    }
  }, [visible]);

  // Remove direct should also check the scroll bar update
  useEffect(
    () => () => {
      switchScrollingEffect();
      clearTimeout(contentTimeoutRef.current);
    },
    [],
  );

  // ========================= Render =========================
  return (
    <div className={`${prefixCls}-root`}>
      <Mask
        prefixCls={prefixCls}
        visible={mask && visible}
        motionName={getMotionName(prefixCls, maskTransitionName, maskAnimation)}
        style={{
          zIndex,
          ...maskStyle,
        }}
        maskProps={maskProps}
      />
      <div
        tabIndex={-1}
        onKeyDown={onWrapperKeyDown}
        className={classNames(`${prefixCls}-wrap`, wrapClassName)}
        ref={wrapperRef}
        onClick={onWrapperClick}
        role="dialog"
        aria-labelledby={title ? ariaIdRef.current : null}
        style={{ zIndex, ...wrapStyle, display: !animatedVisible ? 'none' : null }}
        {...wrapProps}
      >
        <Content
          {...props}
          onClick={onContentClick}
          ref={contentRef}
          closable={closable}
          ariaId={ariaIdRef.current}
          prefixCls={prefixCls}
          visible={visible}
          onClose={onInternalClose}
          onVisibleChanged={onDialogVisibleChanged}
          motionName={getMotionName(prefixCls, transitionName, animation)}
        />
      </div>
    </div>
  );
}
