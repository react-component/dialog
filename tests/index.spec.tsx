/* eslint-disable react/no-render-return-value, max-classes-per-file, func-names, no-console */
import React, { cloneElement } from 'react';
import { act } from 'react-dom/test-utils';
import type { ReactWrapper } from 'enzyme';
import { mount } from 'enzyme';
import Portal from 'rc-util/lib/Portal';
import KeyCode from 'rc-util/lib/KeyCode';
import type { DialogProps } from '../src';
import Dialog from '../src';

describe('dialog', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('show', () => {
    const wrapper = mount(<Dialog visible />);
    jest.runAllTimers();
    wrapper.update();
    expect(wrapper.find('.rc-dialog-wrap').props().style.display).toBeFalsy();
  });

  it('close', () => {
    const wrapper = mount(<Dialog visible />);
    jest.runAllTimers();

    wrapper.setProps({ visible: false });
    jest.runAllTimers();
    wrapper.update();

    expect(wrapper.find('.rc-dialog-wrap').props().style.display).toEqual('none');
  });

  it('create & root & mask', () => {
    const wrapper = mount(<Dialog visible />);
    jest.runAllTimers();
    wrapper.update();

    expect(wrapper.find('.rc-dialog-root').length).toBeTruthy();
    expect(wrapper.find('.rc-dialog-mask').length).toBeTruthy();
  });

  it('click close', () => {
    const onClose = jest.fn();
    const wrapper = mount(<Dialog closeIcon="test" onClose={onClose} visible />);
    jest.runAllTimers();
    wrapper.update();

    const btn = wrapper.find('.rc-dialog-close');
    expect(btn.text()).toBe('test');
    btn.simulate('click');

    jest.runAllTimers();
    wrapper.update();
    expect(onClose).toHaveBeenCalledTimes(1);
  });

  describe('destroyOnClose', () => {
    it('default is false', () => {
      const wrapper = mount(
        <Dialog visible>
          <input className="test-destroy" />
        </Dialog>,
        { attachTo: document.body },
      );

      act(() => {
        wrapper.setProps({ visible: false });
        jest.runAllTimers();
        wrapper.update();
      });

      expect(document.querySelectorAll('.test-destroy')).toHaveLength(1);

      wrapper.unmount();
    });

    it('destroy on hide should unmount child components on close', () => {
      const wrapper = mount(
        <Dialog destroyOnClose>
          <input className="test-input" />
        </Dialog>,
        { attachTo: document.body },
      );

      // Show
      wrapper.setProps({ visible: true });
      jest.runAllTimers();
      wrapper.update();

      ((document.getElementsByClassName('.test-input') as unknown) as HTMLInputElement).value =
        'test';
      expect(
        ((document.getElementsByClassName('.test-input') as unknown) as HTMLInputElement).value,
      ).toBe('test');

      // Hide
      wrapper.setProps({ visible: false });
      jest.runAllTimers();
      wrapper.update();

      // Show
      wrapper.setProps({ visible: true });
      jest.runAllTimers();
      wrapper.update();

      expect(
        ((document.getElementsByClassName('.test-input') as unknown) as HTMLInputElement).value,
      ).toBeUndefined();
      wrapper.unmount();
    });
  });

  it('esc to close', () => {
    const onClose = jest.fn();
    const wrapper = mount(<Dialog onClose={onClose} visible />);
    jest.runAllTimers();
    wrapper.update();

    wrapper.find('.rc-dialog').simulate('keyDown', { keyCode: KeyCode.ESC });
    jest.runAllTimers();
    wrapper.update();
    expect(onClose).toHaveBeenCalled();
  });

  it('mask to close', () => {
    const onClose = jest.fn();
    const wrapper = mount(<Dialog onClose={onClose} visible />);

    // Mask close
    wrapper.find('.rc-dialog-wrap').simulate('click');
    jest.runAllTimers();
    wrapper.update();
    expect(onClose).toHaveBeenCalled();
    onClose.mockReset();

    // Mask can not close
    wrapper.setProps({ maskClosable: false });
    wrapper.find('.rc-dialog-wrap').simulate('click');
    jest.runAllTimers();
    wrapper.update();
    expect(onClose).not.toHaveBeenCalled();
  });

  it('renderToBody', () => {
    const container = document.createElement('div');
    document.body.appendChild(container);
    const wrapper = mount(
      <Dialog visible={false}>
        <p className="renderToBody">1</p>
      </Dialog>,
      { attachTo: container },
    );

    expect(wrapper.find('.renderToBody')).toHaveLength(0);
    expect(wrapper.find('.rc-dialog-wrap')).toHaveLength(0);

    // Visible
    wrapper.setProps({ visible: true });
    jest.runAllTimers();
    wrapper.update();

    expect(wrapper.find('.rc-dialog-wrap')).toHaveLength(1);
    expect(wrapper.find('.renderToBody')).toHaveLength(1);
    expect(container.contains(wrapper.find('.rc-dialog-wrap').getDOMNode())).toBeFalsy();

    wrapper.unmount();
    document.body.removeChild(container);
  });

  it('getContainer', () => {
    const returnedContainer = document.createElement('div');
    const wrapper = mount(
      <Dialog visible getContainer={() => returnedContainer}>
        <p className="getContainer">Hello world!</p>
      </Dialog>,
    );

    expect(returnedContainer.contains(wrapper.find('.rc-dialog-wrap').getDOMNode())).toBeTruthy();
    wrapper.unmount();
  });

  it('render title correctly', () => {
    const wrapper = mount(<Dialog visible title="bamboo" />);
    expect(wrapper.find('.rc-dialog-header').text()).toBe('bamboo');
  });

  it('render footer correctly', () => {
    const wrapper = mount(<Dialog visible footer="test" />);
    expect(wrapper.find('.rc-dialog-footer').text()).toBe('test');
  });

  it('support input autoFocus', () => {
    const wrapper = mount(
      <Dialog visible>
        <input autoFocus />
      </Dialog>,
      { attachTo: document.body },
    );
    expect(document.activeElement).toBe(document.querySelector('input'));
    wrapper.unmount();
  });

  describe('Tab should keep focus in dialog', () => {
    it('basic tabbing', () => {
      const wrapper = mount(<Dialog visible />, { attachTo: document.body });
      const sentinelEnd = (document.querySelectorAll(
        '.rc-dialog-content + div',
      )[0] as unknown) as HTMLDivElement;
      sentinelEnd.focus();

      wrapper.find('.rc-dialog-wrap').simulate('keyDown', {
        keyCode: KeyCode.TAB,
      });

      const sentinelStart = document.querySelectorAll('.rc-dialog > div')[0];
      expect(document.activeElement).toBe(sentinelStart);

      wrapper.unmount();
    });

    it('trap focus after shift-tabbing', () => {
      const wrapper = mount(<Dialog visible />, { attachTo: document.body });
      wrapper.find('.rc-dialog-wrap').simulate('keyDown', {
        keyCode: KeyCode.TAB,
        shiftKey: true,
      });
      const sentinelEnd = document.querySelectorAll('.rc-dialog-content + div')[0];
      expect(document.activeElement).toBe(sentinelEnd);

      wrapper.unmount();
    });
  });

  it('sets transform-origin when property mousePosition is set', () => {
    const wrapper = mount(
      <Dialog style={{ width: 600 }} mousePosition={{ x: 100, y: 100 }} visible>
        <p>the dialog</p>
      </Dialog>,
    );

    // Trigger position align
    act(() => {
      wrapper
        .find<any>('Content CSSMotion' as any)
        .props()
        .onAppearPrepare();
    });

    expect(
      (wrapper.find('.rc-dialog').at(0).getDOMNode() as HTMLDivElement).style['transform-origin'],
    ).toBeTruthy();
  });

  it('can get dom element before dialog first show when forceRender is set true ', () => {
    const wrapper = mount(
      <Dialog forceRender>
        <div>forceRender element</div>
      </Dialog>,
    );
    expect(wrapper.find('.rc-dialog-body > div').text()).toEqual('forceRender element');
  });

  describe('getContainer is false', () => {
    it('not set', () => {
      const wrapper = mount(
        <Dialog visible>
          <div>forceRender element</div>
        </Dialog>,
      );
      expect(wrapper.find('.rc-dialog-body > div').text()).toEqual('forceRender element');
      expect(wrapper.find(Portal)).toHaveLength(1);
    });

    it('set to false', () => {
      const wrapper = mount(
        <Dialog visible getContainer={false}>
          <div>forceRender element</div>
        </Dialog>,
      );
      expect(wrapper.find('.rc-dialog-body > div').text()).toEqual('forceRender element');
      expect(wrapper.find(Portal)).toHaveLength(0);
    });
  });

  it('should not close if mouse down in dialog', () => {
    const onClose = jest.fn();
    const wrapper = mount(<Dialog onClose={onClose} visible />);
    wrapper.find('.rc-dialog-body').simulate('click');
    expect(onClose).not.toHaveBeenCalled();
  });

  it('zIndex', () => {
    const wrapper = mount(<Dialog visible zIndex={903} />);
    expect(wrapper.find('.rc-dialog-wrap').props().style.zIndex).toBe(903);
  });

  it('should show dialog when initialize dialog, given forceRender and visible is true', () => {
    class DialogWrapTest extends React.Component {
      state = {
        visible: true,
        forceRender: true,
      };

      render() {
        return <Dialog forceRender={this.state.forceRender} visible={this.state.visible} />;
      }
    }

    const wrapper = mount(
      <DialogWrapTest>
        <div>Show dialog with forceRender and visible is true</div>
      </DialogWrapTest>,
    );
    jest.runAllTimers();
    wrapper.update();
    expect(wrapper.find('.rc-dialog-wrap').props().style.display).toEqual(null);
  });

  it('modalRender', () => {
    const modalRender = mount(
      <Dialog
        visible
        modalRender={(node: React.ReactElement) =>
          cloneElement(node, { ...node.props, style: { background: '#1890ff' } })
        }
      />,
    );
    expect(modalRender.find('.rc-dialog-content').props().style.background).toEqual('#1890ff');
  });

  describe('size should work', () => {
    it('width', () => {
      const wrapper = mount(<Dialog visible width={1128} />);
      expect(wrapper.find('.rc-dialog').props().style.width).toEqual(1128);
    });

    it('height', () => {
      const wrapper = mount(<Dialog visible height={903} />);
      expect(wrapper.find('.rc-dialog').props().style.height).toEqual(903);
    });
  });

  describe('re-render', () => {
    function createWrapper(props?: Partial<DialogProps>): [ReactWrapper, () => number] {
      let renderTimes = 0;
      const RenderChecker = () => {
        renderTimes += 1;
        return null;
      };

      const Demo = (demoProps?: any) => {
        return (
          <Dialog visible {...props} {...demoProps}>
            <RenderChecker />
          </Dialog>
        );
      };

      const wrapper = mount(<Demo />);

      return [wrapper, () => renderTimes];
    }

    it('should not re-render when visible changed', () => {
      const [wrapper, getRenderTimes] = createWrapper();
      expect(getRenderTimes()).toEqual(1);

      // Hidden should not trigger render
      wrapper.setProps({ visible: false });
      expect(getRenderTimes()).toEqual(1);
    });

    it('should re-render when forceRender', () => {
      const [wrapper, getRenderTimes] = createWrapper({ forceRender: true });
      expect(getRenderTimes()).toEqual(1);

      // Hidden should not trigger render
      wrapper.setProps({ visible: false });
      expect(getRenderTimes()).toEqual(2);
    });
  });

  describe('afterClose', () => {
    it('should trigger afterClose when set visible to false', () => {
      const afterClose = jest.fn();

      const wrapper = mount(<Dialog afterClose={afterClose} visible />);
      jest.runAllTimers();

      wrapper.setProps({ visible: false });
      jest.runAllTimers();

      expect(afterClose).toHaveBeenCalledTimes(1);
    });

    it('should not trigger afterClose when mount dialog of getContainer={false}', () => {
      const afterClose = jest.fn();

      const wrapper = mount(<Dialog afterClose={afterClose} getContainer={false} />);
      jest.runAllTimers();

      wrapper.setProps({ visible: false });
      jest.runAllTimers();

      expect(afterClose).toHaveBeenCalledTimes(0);
    });

    it('should not trigger afterClose when mount dialog of forceRender={true}', () => {
      const afterClose = jest.fn();

      const wrapper = mount(<Dialog afterClose={afterClose} forceRender />);
      jest.runAllTimers();

      wrapper.setProps({ visible: false });
      jest.runAllTimers();

      expect(afterClose).toHaveBeenCalledTimes(0);
    });
  });
});
