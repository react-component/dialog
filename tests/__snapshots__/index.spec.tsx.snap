// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dialog add rootClassName should render correct 1`] = `
<div
  class="rc-dialog-root customize-root-class"
>
  <div
    class="rc-dialog-mask"
  />
  <div
    class="rc-dialog-wrap"
    style="font-size: 10px;"
    tabindex="-1"
  >
    <div
      aria-modal="true"
      class="rc-dialog"
      role="dialog"
      style="width: 600px; height: 903px;"
    >
      <div
        aria-hidden="true"
        style="width: 0px; height: 0px; overflow: hidden; outline: none;"
        tabindex="0"
      />
      <div
        class="rc-dialog-content"
      >
        <button
          aria-label="Close"
          class="rc-dialog-close"
          type="button"
        >
          <span
            class="rc-dialog-close-x"
          />
        </button>
        <div
          class="rc-dialog-body"
        />
      </div>
      <div
        aria-hidden="true"
        style="width: 0px; height: 0px; overflow: hidden; outline: none;"
        tabindex="0"
      />
    </div>
  </div>
</div>
`;

exports[`dialog should render correct 1`] = `
<div
  class="rc-dialog-root"
>
  <div
    class="rc-dialog-mask"
  />
  <div
    class="rc-dialog-wrap"
    tabindex="-1"
  >
    <div
      aria-labelledby="test-id"
      aria-modal="true"
      class="rc-dialog"
      role="dialog"
    >
      <div
        aria-hidden="true"
        style="width: 0px; height: 0px; overflow: hidden; outline: none;"
        tabindex="0"
      />
      <div
        class="rc-dialog-content"
      >
        <button
          aria-label="Close"
          class="rc-dialog-close"
          type="button"
        >
          <span
            class="rc-dialog-close-x"
          />
        </button>
        <div
          class="rc-dialog-header"
        >
          <div
            class="rc-dialog-title"
            id="test-id"
          >
            Default
          </div>
        </div>
        <div
          class="rc-dialog-body"
        />
      </div>
      <div
        aria-hidden="true"
        style="width: 0px; height: 0px; overflow: hidden; outline: none;"
        tabindex="0"
      />
    </div>
  </div>
</div>
`;
